File: backend/services/post-ts-koa-mongo-fdm/src/user-register/impl.ts
import { UserName, Password, UserId } from "../common";
import { Deps, now } from "../common/utils";
import { Workflow } from ".";
import * as db from "../common/utils/db";
import { Long, MongoError } from "mongodb";
import { errors } from "./api";

export class WorkflowImpl extends Workflow {
    async createUser(userName: UserName, password: Password): Promise<UserId> {
        try {
            const rec = await this.deps.mongoDb.collection(db.users).insertOne({
                name: userName,
                encryptedPassword: await this.deps.encryption.encrypt(password.plain),
                creationTime: Long.fromNumber(now().utc),
            } satisfies {
                name: string,
                encryptedPassword: string,
                creationTime: Long,
            });
            return rec.insertedId as UserId;
        } catch (e) {
            if (e instanceof MongoError) {
                if (e.code === 11000) {
                    throw this.errors.userNameAlreadyExists();
                }
            }
            throw e;
        }
    }

    readonly errors = errors;
    constructor(private readonly deps: Deps) { super(); }
}

File: backend/services/post-ktor-mongo/src/main/kotlin/me/jason5lee/post_ktor_mongo_fdm/user_register/impl.kt
package me.jason5lee.post_ktor_mongo.user_register

import com.mongodb.MongoWriteException
import kotlinx.coroutines.reactive.awaitSingle
import me.jason5lee.post_ktor_mongo.common.Password
import me.jason5lee.post_ktor_mongo.common.UserId
import me.jason5lee.post_ktor_mongo.common.UserName
import me.jason5lee.post_ktor_mongo.common.utils.Db
import me.jason5lee.post_ktor_mongo.common.utils.Deps
import me.jason5lee.post_ktor_mongo.common.utils.timeFromNow
import org.bson.Document

class WorkflowImpl(private val deps: Deps) : Workflow(), ErrorsImpl {
    override suspend fun createUser(userName: UserName, password: Password): UserId {
        val doc = Document()
        doc["name"] = userName.value
        doc["encryptedPassword"] = password.encrypt(deps.bCrypt)
        doc["creationTime"] = timeFromNow(offset = 0).utc

        val inserted = try {
            deps.mongoDb.getCollection(Db.users).insertOne(doc).awaitSingle()
        } catch (e: MongoWriteException) {
            throw userNameAlreadyExists()
        }

        return UserId(Db.formatId(inserted.insertedId!!.asObjectId().value))
    }
}
